Ldap federation seems to open and keep open a new thread/connection for each ldap request

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

ldap

### Describe the bug

Keycloak seems to open and keep open for a while a new connection to the ldap server for each request. At least in below scenario:

I'm using the keycloak-admin-client 26.0.4 witch keycloak 26.1.4.
There is a writable Ldap User Federation set up with openLdap 2.5.13+dfsg-5

I access all users doing basically that here.
Nothing will be written back to keycloak. It's only read access

```
for(var user: allUsers) {
    List<UserRepresentation> users = midwifeRealm.users().search(user.getUid(), true);
    if (users.isEmpty()) {
         // throw exception
    }
    userRepresentation = user.get(0);
    // do something 
    // or do nothing. for testing I've just run through the loop.
}
```
At one point the users.isEmpty() call fails with:
```
Apr 03 00:01:48 intranet java[820274]: jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request
Apr 03 00:01:48 intranet java[820274]:         at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:236)
Apr 03 00:01:48 intranet java[820274]:         at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:216)
Apr 03 00:01:48 intranet java[820274]:         at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.BodyEntityExtractor.extractEntity(BodyEntityExtractor.java:59)
Apr 03 00:01:48 intranet java[820274]:         at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:136)
Apr 03 00:01:48 intranet java[820274]:         at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103)
Apr 03 00:01:48 intranet java[820274]:         at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:102)
Apr 03 00:01:48 intranet java[820274]:         at jdk.proxy2/jdk.proxy2.$Proxy279.search(Unknown Source)
Apr 03 00:01:48 intranet java[820274]:         at midwives.hebammenprofil.service.impl.KeycloakService.getUser(KeycloakService.java:152)

```
and openldap logs:
`slapd[2974]: warning: cannot open /etc/hosts.deny: Too many open files`

using pstree I could see that the keycloak container had around 1000 threads running.
Thread {Thread-7680} to {Thread-8713}. I didn't check if any thread number was missing, but they seemed to be sequential.
Even some minutes after the exception the threads where still there, and new requests failed. I didn't check how long it would take to everything go back to normal. At one point it does, since we've got a job running during night, which triggers that problem, but in the morning everything is back to normal. So in normal usage that problem doesn't appear.

After restarting openLdap all those Keycloak threads disappear and everything goes back to normal.

I checked if UserRepresentation is a resource that should be closed after using it. But couldn't find a close() possibility.

So as it seems keycloak starts up a new thread for each users.isEmpty() and/or users().search and/or users.get(0) where it access openldap. And keeps the thread and the connection to the ldap server open for a while instead of reusing the connection or just dropping it.

Is there a good reason to do so?
Is there a misconfiguration on my side?
Would it be possible to get a better exception than "jakarta.ws.rs.BadRequestException: HTTP 400 Bad Request"? It took me quite a while to realise, that the problem is not a badly formed request.


### Version

26.1.4

### Regression

- [ ] The issue is a regression

### Expected behavior

To close or reuse the ldap connection.

### Actual behavior

Keeping them open for a while

### How to Reproduce?

see above

### Anything else?

ldap config:

![Image](https://github.com/user-attachments/assets/62ce3562-b0a3-4ed7-b373-37580653a65f)
![Image](https://github.com/user-attachments/assets/5bb74bf4-89b3-4113-bd64-4f23da652aee)
![Image](https://github.com/user-attachments/assets/a1d9ee37-6b0b-43d6-84f7-5188de28f9e9)

just noticed, after maybe 20 minutes keycloak throws following exception.

```
2025-04-11 16:46:11,495 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012108: CheckedAction::check - atomic action 0:ffffac110003:98c7:67f944d2:186 aborting with 1 threads active!
2025-04-11 16:46:11,497 WARN  [org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorTrackingImpl] (Transaction Reaper Worker 0) HHH000451: Transaction afterCompletion called by a background thread; delaying afterCompletion processing until the original thread can handle it. [status=4]
2025-04-11 16:46:11,497 WARN  [com.arjuna.ats.arjuna] (Transaction Reaper Worker 0) ARJUNA012121: TransactionReaper::doCancellations worker Thread[#36,Transaction Reaper Worker 0,5,main] successfully canceled TX 0:ffffac110003:98c7:67f944d2:186
2025-04-11 16:50:34,275 WARN  [com.arjuna.ats.arjuna] (executor-thread-2) ARJUNA012077: Abort called on already aborted atomic action 0:ffffac110003:98c7:67f944d2:186
2025-04-11 16:50:34,275 WARN  [com.arjuna.ats.arjuna] (executor-thread-1) ARJUNA012077: Abort called on already aborted atomic action 0:ffffac110003:98c7:67f944d2:168
2025-04-11 16:50:34,283 WARN  [org.keycloak.services.DefaultKeycloakSession] (executor-thread-2) Unable to close provider org.keycloak.connections.jpa.DefaultJpaConnectionProvider: org.keycloak.models.ModelException: Database operation failed
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.convert(PersistenceExceptionConverter.java:106)
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:68)
	at jdk.proxy2/jdk.proxy2.$Proxy53.close(Unknown Source)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProvider.close(DefaultJpaConnectionProvider.java:44)
	at org.keycloak.services.DefaultKeycloakSession.lambda$close$5(DefaultKeycloakSession.java:367)
	at java.base/java.util.HashMap$Values.forEach(HashMap.java:1073)
	at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:373)
	at org.keycloak.models.KeycloakBeanProducer_ProducerMethod_getKeycloakSession_XoSEUTXOsE3bpqXlGMAykCiECUM_ClientProxy.close(Unknown Source)
	at org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler.close(TransactionalSessionHandler.java:60)
	at org.keycloak.quarkus.runtime.integration.jaxrs.CloseSessionFilter.closeSession(CloseSessionFilter.java:67)
	at java.base/java.util.stream.Streams$1.run(Streams.java:858)
	at java.base/java.util.stream.AbstractPipeline.close(AbstractPipeline.java:323)
	at com.fasterxml.jackson.datatype.jdk8.StreamSerializer.serialize(StreamSerializer.java:84)
	at com.fasterxml.jackson.datatype.jdk8.StreamSerializer.serialize(StreamSerializer.java:15)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:502)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:422)
	at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1579)
	at com.fasterxml.jackson.databind.ObjectWriter._writeValueAndClose(ObjectWriter.java:1285)
	at com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:1098)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyWriter.writeResponse(FullyFeaturedServerJacksonMessageBodyWriter.java:106)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyWriter_ClientProxy.writeResponse(Unknown Source)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:186)
	at org.jboss.resteasy.reactive.server.core.serialization.FixedEntityWriter.write(FixedEntityWriter.java:28)
	at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:34)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:147)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:638)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.hibernate.HibernateException: Transaction was rolled back in a different thread
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorTrackingImpl.processAnyDelayedAfterCompletion(SynchronizationCallbackCoordinatorTrackingImpl.java:90)
	at org.hibernate.internal.SessionImpl.delayedAfterCompletion(SessionImpl.java:611)
	at org.hibernate.internal.AbstractSharedSessionContract.close(AbstractSharedSessionContract.java:464)
	at org.hibernate.internal.SessionImpl.closeWithoutOpenChecks(SessionImpl.java:414)
	at org.hibernate.internal.SessionImpl.close(SessionImpl.java:395)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:66)
	... 34 more

2025-04-11 16:50:34,283 WARN  [org.keycloak.services.DefaultKeycloakSession] (executor-thread-1) Unable to close provider org.keycloak.connections.jpa.DefaultJpaConnectionProvider: org.keycloak.models.ModelException: Database operation failed
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.convert(PersistenceExceptionConverter.java:106)
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:68)
	at jdk.proxy2/jdk.proxy2.$Proxy53.close(Unknown Source)
	at org.keycloak.connections.jpa.DefaultJpaConnectionProvider.close(DefaultJpaConnectionProvider.java:44)
	at org.keycloak.services.DefaultKeycloakSession.lambda$close$5(DefaultKeycloakSession.java:367)
	at java.base/java.util.HashMap$Values.forEach(HashMap.java:1073)
	at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:373)
	at org.keycloak.models.KeycloakBeanProducer_ProducerMethod_getKeycloakSession_XoSEUTXOsE3bpqXlGMAykCiECUM_ClientProxy.close(Unknown Source)
	at org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler.close(TransactionalSessionHandler.java:60)
	at org.keycloak.quarkus.runtime.integration.jaxrs.CloseSessionFilter.closeSession(CloseSessionFilter.java:67)
	at java.base/java.util.stream.Streams$1.run(Streams.java:858)
	at java.base/java.util.stream.AbstractPipeline.close(AbstractPipeline.java:323)
	at com.fasterxml.jackson.datatype.jdk8.StreamSerializer.serialize(StreamSerializer.java:84)
	at com.fasterxml.jackson.datatype.jdk8.StreamSerializer.serialize(StreamSerializer.java:15)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:502)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:422)
	at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1579)
	at com.fasterxml.jackson.databind.ObjectWriter._writeValueAndClose(ObjectWriter.java:1285)
	at com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:1098)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyWriter.writeResponse(FullyFeaturedServerJacksonMessageBodyWriter.java:106)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyWriter_ClientProxy.writeResponse(Unknown Source)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
	at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:186)
	at org.jboss.resteasy.reactive.server.core.serialization.FixedEntityWriter.write(FixedEntityWriter.java:28)
	at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:34)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:147)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:638)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.hibernate.HibernateException: Transaction was rolled back in a different thread
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorTrackingImpl.processAnyDelayedAfterCompletion(SynchronizationCallbackCoordinatorTrackingImpl.java:90)
	at org.hibernate.internal.SessionImpl.delayedAfterCompletion(SessionImpl.java:611)
	at org.hibernate.internal.AbstractSharedSessionContract.close(AbstractSharedSessionContract.java:464)
	at org.hibernate.internal.SessionImpl.closeWithoutOpenChecks(SessionImpl.java:414)
	at org.hibernate.internal.SessionImpl.close(SessionImpl.java:395)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:66)
	... 34 more

```

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/store/ldap/LDAPContextManager.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/store/ldap/LDAPContextManager.java
@@ -11,7 +11,6 @@ import org.keycloak.vault.VaultStringSecret;
 import javax.naming.AuthenticationException;
 import javax.naming.Context;
 import javax.naming.NamingException;
-import javax.naming.ldap.InitialLdapContext;
 import javax.naming.ldap.LdapContext;
 import javax.naming.ldap.StartTlsRequest;
 import javax.naming.ldap.StartTlsResponse;
@@ -78,7 +77,7 @@ public final class LDAPContextManager implements AutoCloseable {
                 connProp.put(LDAPConstants.CONNECTION_TRACE_BER, System.err);
             }
 
-            ldapContext = new InitialLdapContext(connProp, null);
+            ldapContext = new SessionBoundInitialLdapContext(session, connProp, null);
             if (ldapConfig.isStartTls()) {
                 SSLSocketFactory sslSocketFactory = null;
                 if (LDAPUtil.shouldUseTruststoreSpi(ldapConfig)) {

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/store/ldap/SessionBoundInitialLdapContext.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/store/ldap/SessionBoundInitialLdapContext.java
@@ -1,0 +1,47 @@
+/*
+ * Copyright 2025 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.storage.ldap.idm.store.ldap;
+
+import org.keycloak.models.KeycloakSession;
+
+import javax.naming.NamingException;
+import javax.naming.ldap.Control;
+import javax.naming.ldap.InitialLdapContext;
+import java.util.Hashtable;
+
+/**
+ * A {@link InitialLdapContext} that binds instances of this class with the {@link KeycloakSession} so that any resource
+ * acquired during the session lifetime is closed when the session is closed.
+ */
+public final class SessionBoundInitialLdapContext extends InitialLdapContext {
+
+    public SessionBoundInitialLdapContext(KeycloakSession session, Hashtable<?, ?> environment, Control[] connCtls) throws NamingException {
+        super(environment, connCtls);
+        session.enlistForClose(() -> {
+                try {
+                    close();
+                } catch (NamingException e) {
+                    failedToCloseLdapContext(e);
+                }
+        });
+    }
+
+    private void failedToCloseLdapContext(NamingException e) {
+        throw new RuntimeException("Failed to close LDAP context", e);
+    }
+}

